pkg/subscriptions/insights.jsx

```diff
@@ -0,0 +1,330 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React from "react";
+import moment from "moment";
+import * as python from "python.js";
+import { show_modal_dialog } from "cockpit-components-dialog.jsx";
+import * as service from "service.js";
+
+import { client } from "./subscriptions-client";
+
+import get_monotonic_start_py from "raw-loader!./get_monotonic_start.py";
+
+let _ = cockpit.gettext;
+
+moment.locale(cockpit.language);
+
+const insights_timer = service.proxy("insights-client.timer", "Timer");
+const insights_service = service.proxy("insights-client.service", "Service");
+
+export function detect() {
+    return cockpit.spawn([ "which", "insights-client" ], { err: "ignore" }).then(() => true, () => false);
+}
+
+function catch_error(err) {
+    let msg = err.toString();
+    // The insights-client frequently dumps
+    // Python backtraces on us. Make them more
+    // readable by wrapping the text in <pre>.
+    if (msg.indexOf("\n") > 0)
+        msg = <pre>{msg}</pre>;
+    client.setError("error", msg);
+}
+
+export function register() {
+    return cockpit.spawn([ "insights-client", "--register" ], { superuser: true, err: "message" })
+            .catch(catch_error);
+}
+
+export function unregister() {
+    if (insights_timer.enabled) {
+        return cockpit.spawn([ "insights-client", "--unregister" ], { superuser: true, err: "message" })
+                .catch(catch_error);
+    } else {
+        return cockpit.resolve();
+    }
+}
+
+// TODO - generalize this to arbitrary number of arguments (when needed)
+export function fmt_to_fragments(fmt, arg) {
+    var index = fmt.indexOf("$0");
+    if (index >= 0)
+        return <React.Fragment>{fmt.slice(0, index)}{arg}{fmt.slice(index + 2)}</React.Fragment>;
```

Comment:
Not important here I figure, but in master we use `<>` instead of `<React.Fragment>` due to new eslint rules.

Reply:
Intersting, I didn't know about `<>`.